### ISSUE TITLE
Unable to find the "ojs_define" function in child qmd after v1.3.433

### ISSUE THREAD

**Issue by Eugloh** [quarto-dev/quarto-cli#6141]
### Bug description

First of all, I'd like to thank you for Quarto and the features you've provided for the automated reports I use on a daily basis.

I'm currently setting up an interface to generate automated HTML analysis reports, such as a quality report.

My idea is to call .qmd children from a file according to the parameters specified in the interface.

but since I encountered the error [#5762](https://github.com/quarto-dev/quarto-cli/issues/5762) and installed v1.3.433 I can no longer generate the OJS code for my report, "ojs_define" function is not found.

If I make it simpler and replace the call child chunk with the code present in q2.qmd the ojs is executed, but it won't be displayed when I reload my html file or try to move it.

## p1.qmd
  ````
  ---
  params:
    analyse_type: "Quality Control"
  format:
    html:
      code-fold: TRUE
      embed-resources: TRUE
  title: |
    Analysis Type : `r params$analyse_type` Report
  cache: FALSE
  ---

  ```{r}
  nb_patients <- 15
  nb_clusters <- 7
  set.seed(1234)
  abundance <- matrix(rnorm(nb_patients * nb_clusters), nrow = nb_patients, ncol = nb_clusters)
  ```

  ```{r , child="p2.qmd", eval =  {if(params$analyse_type=="Quality Control")TRUE else FALSE}  } 
  
  ```

````

## p2.qmd
  ````
  # Explore PCA


  ```{r}
  abundance |> as.data.frame() |>
    dplyr::mutate_all(scale, center = TRUE) -> centered
  pca <- stats::prcomp(centered) 
  Table <- as.data.frame(predict(pca))
  ojs_define(pcjs = Table)

  ```

  ::: {.panel-sidebar}
  ```{ojs}

  ojs_pc = transpose(pcjs)
  viewof pc1 = Inputs.select(Object.keys(ojs_pc[0]), {value: "PC1", multiple: false, label: "PC Axe 1"})
  viewof pc2 = Inputs.select(Object.keys(ojs_pc[0]), {value: "PC2", multiple: false, label: "PC Axe 2"})

  ```
  :::

  ::: {.panel-fill}
  ::: {.panel-tabset}
  ## Plot

  ```{ojs}
  //| label: fig-pca-with-ojs
  //| fig-cap: "Interactive pca plot"
  //| panel: fill

  Plot.plot({
    grid: true,
    inset: 10,
    x: {label: pc1},
    y: {label: pc2},
    marks: [
    Plot.frame(),
      Plot.dot(ojs_pc, {x: pc1, y: pc2})
    ]
  })
  ```

  ## Data

  ```{ojs}
  Inputs.table(ojs_pc)
  ```
  :::
  :::
  ````

### Expected behavior

Previously, the ojs_define function worked even when it was in a child file, and the ojs output was self-contained in the html file.

### Actual behavior

I am now getting :
```
processing file: p1.qmd
1/5                  
2/5 [unnamed-chunk-1]
3/5                  
4/5 [unnamed-chunk-2]


processing file: <path>/p2.qmd
1/9                  
2/9 [unnamed-chunk-3]



Quitting from lines 6-1 [unnamed-chunk-3] (<path>/p2.qmd)



Quitting from lines 1-19 [unnamed-chunk-2] (<path>/p2.qmd)
Error in `ojs_define()`:
! impossible de trouver la fonction "ojs_define"
Exécution arrêtée
```

### Your environment

**R version 4.2.2 Patched (2022-11-10 r83330)**

**Platform:** x86_64-pc-linux-gnu (64-bit) 

**RStudio:** 2023.06.0 Build 421

### Quarto check output

quarto check
```
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.2: OK
      Dart Sass version 1.55.0: OK
      Deno version 1.33.2: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.4.193
      Path: /opt/quarto/bin

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.10.6
      Path: /usr/bin/python3
      Jupyter: 4.9.1
      Kernels: julia-1.6, python3

(|) Checking Jupyter engine render....[IPKernelApp] WARNING | debugpy_stream undefined, debugging will not be enabled
[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.2.2
      Path: /usr/lib/R
      LibPaths:
        - /usr/local/lib/R/site-library
        - /usr/lib/R/site-library
        - /usr/lib/R/library
      knitr: 1.43
      rmarkdown: 2.22

[✓] Checking Knitr engine render......OK
```

**Comment by mcanouil** [comment]
This seems to be because you are including a document not using Quarto but using `knitr`.
So the `obj_define()` defined by Quarto is not discoverable in that case.

If you replace the "child" thing (see below the Quarto/yaml syntax, and improved R code):

````qmd
```{r}
#| child: "p2.qmd"
#| eval: !expr 'params$analyse_type == "Quality Control"'
```
````

With the following (yes, here for the sake of the example, you loose the conditional aspect, which you could get back with <https://quarto.org/docs/projects/profiles.html> and <https://quarto.org/docs/authoring/conditional.html>.)

````qmd
{{< include p2.qmd >}}
````

@cderv as you are the one with the best knowledge of `knitr`, do you know if there is something to be done here to expose `obj_define` to all computations?

**Comment by cderv** [comment]
> This seems to be because you are including a document not using Quarto but using knitr.

Just to clarify in case it is not to every reader. `ojs_define()` is working in **knitr** document 

````markdown
---
title: "R and OJS"
format: html
engine: knitr
---

```{r}
x = 3
ojs_define(x_ojs = x)
```

```{ojs}
x_ojs * x_ojs
```
````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/8a2300dc-a9c7-4370-b096-80ae1ec66f89)

but indeed here the `{ojs}` chunk is inside the child document and this child document using `child = ` option will be processed by **knitr** which is not aware of OJS chunk. 

I don't know if this is only a matter of making `ojs_define()` available to **knitr** - the `{ojs}` chunk may needs to be processed at the "first level" meaning not in a knitr child. 

That is why "Quarto child" is better in this case, and `{{< include ... >}}` seems the way to go. 

I'll look into that though to check if there is something missing...

**Comment by cderv** [comment]
> do you know if there is something to be done here to expose obj_define to all computations?

So in fact, we make `ojs_define()` available only when Quarto finds a `ojs` chunk in the document, but this does not look into child doc (which is a knitr thing) 
https://github.com/quarto-dev/quarto-cli/blob/c8fb73c433215926d1b8b18f39e29bc1a215e33d/src/resources/rmd/execute.R#L120-L123

I believe in this case, it is best to use [include shortcode](https://quarto.org/docs/authoring/includes.html) to add the child content. I don't think we want to add specific knowledge of child document as this is knitr handling it. 

So basically `ojs_define()` works in child document for knitr, if there is at least a `ojs` chunk in the main document.

**Comment by Eugloh** [comment]
Indeed, if I declare an ojs chunk in my main file, I no longer get the "ojs_define" error.

Did I understand correctly the alternative of "include" which would be for example : 

```
`r if (params$analyse_type != "QC") "::: {.content-hidden}"`

{{< include QC.qmd >}}

`r if (params$analyse_type != "QC") ":::"`
```

because this forces the reading of all the chunks of each of my templates which is not necessary as they are independent.

Unfortunately I've given up on my ojs chunks for now as I get _'the OJS runtime does not work with file// URLS. Please us a web server to view this document'_

Thanks a lot.

**Comment by cderv** [comment]
> the OJS runtime does not work with file// URLS. Please us a web server to view this document

Viewing OJS document requires to preview them in browser using a webserver. `quarto preview` does it. in R `servr::httpd()` will help. 

> Did I understand correctly the alternative of "include" which would be for example :

Indeed as you are using conditional with params for evaluation this is tricky. 

To avoid changing all to this, you could still use child but add a dummy `ojs` chunk in your main document that would not evaluation or be included 

````markdown
```{ojs}
//| echo: false
//| eval: false
1 +1 
```
````

`ojs` chunk would still be detected, and so `ojs_define()` loaded for the child document to use.
